## **MongoDB NoSQL Database Design and Implementation**

This notebook demonstrates the design and implementation of a MongoDB-based NoSQL database for the NorthStar Urban Mobility and
logistics system. It focuses on modelling operational data such as deliveries, customers, and app sessions using a document-oriented approach. The workflow includes data loading, CRUD operations, aggregation analysis & query optimisation.

**Rationale for Using NoSQL (MongoDB)**

MongoDB is used due to its flexible schema design, which supports complex and evolving logistics data in the NorthStar Urban Mobility and Logistics system. It enables embedding of related entities such as orders, drivers, and vehicles within delivery documents, reducing join complexity. This improves query performance and is suitable for large-scale, real-time operational and analytical workloads.

In [1]:
!pip install pymongo dnspython pandas --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 11.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 331.1/331.1 kB 16.0 MB/s eta 0:00:00


### **Import Libraries**

In [112]:
import pandas as pd
from pymongo import MongoClient
from pymongo import ASCENDING, DESCENDING
from datetime import datetime, timezone

### **Connect to MongoDB Atlas**

In [113]:
MONGO_URI = "mongodb+srv://nehasavahircs_db_user:neha2004@cluster1.spmhjjt.mongodb.net/"

client = MongoClient(MONGO_URI)
db = client["northstar_db"]

print("Connected successfully.")

Connected successfully.


A connection was successfully established with MongoDB Atlas, and a database named northstar_db was created to store all operational collections.

### **Load Data From Github**

In [114]:
base_url = "https://raw.githubusercontent.com/Neha-Savahir/northstar-analysis/main/northstar_dataset/"

deliveries_df = pd.read_csv(base_url + "deliveries.csv")
customers_df = pd.read_csv(base_url + "customers.csv")
drivers_df = pd.read_csv(base_url + "drivers.csv")
vehicles_df = pd.read_csv(base_url + "vehicles.csv")
orders_df = pd.read_csv(base_url + "orders.csv")
hubs_df = pd.read_csv(base_url + "hubs.csv")
complaints_df = pd.read_csv(base_url + "complaints.csv")
incidents_df = pd.read_csv(base_url + "incidents.csv")
app_events_df = pd.read_csv(base_url + "app_events.csv")

In [115]:
col_deliveries.delete_many({})
col_customers.delete_many({})
col_sessions.delete_many({})


DeleteResult({'n': 639, 'electionId': ObjectId('7fffffff000000000000014e'), 'opTime': {'ts': Timestamp(1778586944, 237), 't': 334}, 'ok': 1.0, '$clusterTime': {'clusterTime': Timestamp(1778586944, 237), 'signature': {'hash': b'\xfd\x95[\x0c_\xf4?\xb2\x0c\xab\xfd\x99\xe5\xf9n\xf17\x85t\xc5', 'keyId': 7592293558351036417}}, 'operationTime': Timestamp(1778586944, 237)}, acknowledged=True)

In [116]:
print("Delivery documents inserted:", col_deliveries.count_documents({}))
print("Customers documents inserted:", col_customers.count_documents({}))
print("Session documents inserted:", col_sessions.count_documents({}))

Delivery documents inserted: 0
Customers documents inserted: 0
Session documents inserted: 0


### **Data Standardisation**

In [117]:
# Text cleaning function
def clean_text(x):
    if pd.isna(x):
        return None
    return str(x).strip().title()


orders_df["pickup_zone"] = orders_df["pickup_zone"].apply(clean_text)
orders_df["dropoff_zone"] = orders_df["dropoff_zone"].apply(clean_text)

hubs_df["zone"] = hubs_df["zone"].apply(clean_text)

customers_df["home_zone"] = customers_df["home_zone"].apply(clean_text)

app_events_df["zone_context"] = app_events_df["zone_context"].apply(clean_text)


deliveries_df["delivery_status"] = deliveries_df["delivery_status"].str.strip().str.lower()

Interpretation:

Zone-related fields were standardized using consistent text formatting to ensure uniformity across datasets. This helps prevent mismatches during joins and aggregation (e.g., “CENTRAL” vs “Central”).

# **MongoDB Schema Design**

###**Define Collections**

In [118]:
col_deliveries = db["delivery_cases"]
col_customers = db["customer_profiles"]
col_sessions = db["app_sessions"]

**Interpretation:**

Three main collections are created to represent different business domains: delivery operations, customer profiles, and app session logs. This forms the base structure of the NoSQL database.

## **Schema Design Concept**

The MongoDB database for the NorthStar Urban Mobility and Logistics system follows a hybrid schema design approach, combining both embedded and referenced data models to balance performance and data consistency.

* Embedded documents are used in delivery cases to store closely related operational data such as orders, drivers, vehicles, hubs, incidents, and complaints within a single document. This improves query efficiency by reducing the need for joins and enabling fast retrieval of complete delivery context.

* Referenced fields are used for reusable entities such as customer_id, driver_id, and hub_id, which appear across multiple records. This ensures data consistency, avoids duplication, and supports scalable relationships across collections.

This hybrid approach optimizes the database for both fast analytical queries and maintainable data structure in large-scale logistics operations.

### **Build Delivery Cases**

In [119]:
# Safe DateTime Function
def safe_datetime(value):
    dt = pd.to_datetime(value, errors="coerce")
    if pd.isna(dt):
        return None
    return dt.to_pydatetime()

In [120]:
delivery_docs = []

for _, row in deliveries_df.iterrows():

    order_match = orders_df[orders_df["order_id"] == row["order_id"]]
    driver_match = drivers_df[drivers_df["driver_id"] == row["driver_id"]]
    vehicle_match = vehicles_df[vehicles_df["vehicle_id"] == row["vehicle_id"]]
    hub_match = hubs_df[hubs_df["hub_id"] == row["hub_id"]]
    incident_match = incidents_df[incidents_df["delivery_id"] == row["delivery_id"]]
    complaint_match = complaints_df[complaints_df["order_id"] == row["order_id"]]

    # ORDER
    order_doc = order_match.to_dict("records")[0] if not order_match.empty else {}
    if "order_created_at" in order_doc:
        order_doc["order_created_at"] = safe_datetime(order_doc.get("order_created_at"))

    # DRIVER
    driver_doc = driver_match.to_dict("records")[0] if not driver_match.empty else {}

    # VEHICLE
    vehicle_doc = vehicle_match.to_dict("records")[0] if not vehicle_match.empty else {}
    if "commission_date" in vehicle_doc:
        vehicle_doc["commission_date"] = safe_datetime(vehicle_doc.get("commission_date"))

    # HUB
    hub_doc = hub_match.to_dict("records")[0] if not hub_match.empty else {}

    # INCIDENTS
    incidents_list = incident_match.to_dict("records")
    for i in incidents_list:
        i["reported_at"] = safe_datetime(i.get("reported_at"))

    # COMPLAINTS
    complaints_list = complaint_match.to_dict("records")
    for c in complaints_list:
        if "created_at" in c:
            c["created_at"] = safe_datetime(c.get("created_at"))

    doc = {
        "delivery_id": row["delivery_id"],
        "order_id": row["order_id"],
        "driver_id": row["driver_id"],
        "vehicle_id": row["vehicle_id"],
        "hub_id": row["hub_id"],

        "dispatch_time": safe_datetime(row["dispatch_time"]),
        "completion_time": safe_datetime(row["delivery_completed_at"]),

        "status": str(row["delivery_status"]).strip().lower(),

        "metrics": {
            "distance_km": row["route_distance_km"],
            "manual_override": row["manual_route_override_count"],
            "missing_proof": row["proof_of_completion_missing"],
            "rating": row["customer_rating_post_delivery"],
            "fuel_cost": row["fuel_or_charge_cost"]
        },

        "order": order_doc,
        "driver": driver_doc,
        "vehicle": vehicle_doc,
        "hub": hub_doc,

        "incidents": incidents_list,
        "complaints": complaints_list
    }

    delivery_docs.append(doc)

print("Delivery documents prepared successfully.")

Delivery documents prepared successfully.


Delivery cases are built using embedded document structure, combining order, driver, vehicle, hub, incidents, and complaints into a single document. This reduces query complexity and improves read performance.

### **Build Customer Profiles**

In [121]:
customer_docs = []

for _, row in customers_df.iterrows():

    doc = {
        "customer_id": row["customer_id"],
        "age": int(row["age"]) if pd.notna(row["age"]) else None,
        "home_zone": row["home_zone"],
        "customer_type": row["customer_type"],
        "signup_date": safe_datetime(row["signup_date"]),

        "engagement": {
            "loyalty_score": row["loyalty_score"],
            "app_engagement_score": row["app_engagement_score"],
            "preferred_channel": row["preferred_channel"],
            "account_status": row["account_status"]
        }
    }

    customer_docs.append(doc)

Customer profiles are structured into a single document containing demographic and engagement information. This supports fast retrieval for customer-level analytics.

### **Build App Sessions**

In [122]:
session_docs = []

for _, row in app_events_df.iterrows():

    ts = safe_datetime(row["event_timestamp"])

    doc = {
        "session_id": row["session_id"],
        "customer_id": row["customer_id"],
        "event_type": row["event_type"],
        "timestamp": ts,
        "device_type": row["device_type"],
        "zone_context": row["zone_context"],
        "api_latency_ms": row["api_latency_ms"],
        "success_flag": row["success_flag"]
    }

    session_docs.append(doc)

print("Session documents prepared successfully.")

Session documents prepared successfully.


App session data is stored as event-level documents capturing user interactions, device type, latency, and success flags for behavioural and performance analysis.


# **CRUD Operations**

CRUD operations are the fundamental database operations used to Create, Read, Update, and Delete data within a database system.

# **Insert**
Adds new documents into a MongoDB collection.

### **Insert Delivery Documents**

In [123]:
col_deliveries.insert_many(delivery_docs)

print("Delivery documents inserted:", col_deliveries.count_documents({}))

Delivery documents inserted: 950


### **Insert Customer Profiles**

In [124]:
col_customers.insert_many(customer_docs)

print("Customer documents inserted:", col_customers.count_documents({}))

Customer documents inserted: 650


### **Insert App Sessions (Event Data)**

In [125]:
col_sessions.insert_many(session_docs)

print("Session documents inserted:", col_sessions.count_documents({}))

Session documents inserted: 640


**Interpretation:**

All prepared documents were inserted into MongoDB collections using insert_many(). This step confirms successful transformation from structured CSV data into document-based NoSQL format.

## **Read**
Retrieves and queries documents from a MongoDB collection.

### **Display First 10 Session IDs**

In [126]:
for doc in col_sessions.find({}, {"_id": 0, "session_id": 1}).limit(10):
    print(doc)

{'session_id': 'S19847'}
{'session_id': 'S32766'}
{'session_id': 'S99516'}
{'session_id': 'S41236'}
{'session_id': 'S12030'}
{'session_id': 'S66655'}
{'session_id': 'S79495'}
{'session_id': 'S59028'}
{'session_id': 'S76526'}
{'session_id': 'S24618'}


**Interpretation**

This query retrieves and displays the first 10 session IDs from the app_sessions collection while excluding the default MongoDB _id field. It demonstrates a basic Read operation with field projection and result limiting.

### **Retrieve a Failed** **Delivery Record**

In [127]:
col_deliveries.find_one({"status": "failed"})

{'_id': ObjectId('6a031560952c265b6092f142'),
 'delivery_id': 'DL00001',
 'order_id': 'O00938',
 'driver_id': 'D004',
 'vehicle_id': 'V056',
 'hub_id': 'H05',
 'dispatch_time': datetime.datetime(2024, 6, 18, 10, 57),
 'completion_time': datetime.datetime(2024, 6, 19, 9, 5, 59, 904000),
 'status': 'failed',
 'metrics': {'distance_km': 17.26,
  'manual_override': 1,
  'missing_proof': 0,
  'rating': 3.07,
  'fuel_cost': 12.05},
 'order': {'order_id': 'O00938',
  'customer_id': 'C0567',
  'service_type': 'Business',
  'order_created_at': datetime.datetime(2024, 6, 18, 9, 48),
  'promised_window_hours': 6,
  'pickup_zone': 'Central',
  'dropoff_zone': 'Central',
  'priority_level': 'Medium',
  'order_value': 151.14,
  'booking_channel': 'Web',
  'special_handling_flag': 0},
 'driver': {'driver_id': 'D004',
  'base_zone': 'Airport',
  'employment_type': 'PartTime',
  'years_experience': 13,
  'training_score': 88.9,
  'driver_rating': 4.75,
  'shift_preference': 'Morning',
  'active_flag': 

**Interpretation**

This query retrieves a single failed delivery document from the delivery_cases collection.

### **Retrieve High Latency Application Sessions**

In [128]:
for doc in col_sessions.find(
    {"api_latency_ms": {"$gt": 700}}
).limit(10):
    print(doc)

{'_id': ObjectId('6a031571952c265b6092f9d1'), 'session_id': 'S63193', 'customer_id': 'C0231', 'event_type': 'delivery_instruction_update', 'timestamp': datetime.datetime(2024, 5, 6, 13, 33), 'device_type': 'Android', 'zone_context': 'Airport', 'api_latency_ms': 704, 'success_flag': 1}
{'_id': ObjectId('6a031571952c265b6092f794'), 'session_id': 'S88326', 'customer_id': 'C0179', 'event_type': 'track_order', 'timestamp': datetime.datetime(2025, 3, 31, 17, 14), 'device_type': 'Android', 'zone_context': 'South', 'api_latency_ms': 707, 'success_flag': 1}
{'_id': ObjectId('6a031571952c265b6092f9c6'), 'session_id': 'S36767', 'customer_id': 'C0203', 'event_type': 'chat_opened', 'timestamp': datetime.datetime(2024, 3, 7, 8, 14), 'device_type': 'Android', 'zone_context': 'West', 'api_latency_ms': 708, 'success_flag': 1}
{'_id': ObjectId('6a031571952c265b6092f9b7'), 'session_id': 'S13907', 'customer_id': 'C0538', 'event_type': 'chat_opened', 'timestamp': datetime.datetime(2024, 12, 15, 9, 29), 'de

**Interpretation:**

This query retrieves application session records where API latency exceeds 700 milliseconds. It helps identify performance bottlenecks and slow user interactions within the app_sessions collection, supporting operational monitoring and user experience analysis.

# **Update**
Modifies existing documents in a MongoDB collection.

### **Update Failed Deliveries Priority**

In [129]:
col_deliveries.update_many(
    {"status": "failed"},
    {"$set": {"priority": "high"}}
)

UpdateResult({'n': 132, 'electionId': ObjectId('7fffffff000000000000014e'), 'opTime': {'ts': Timestamp(1778587005, 133), 't': 334}, 'nModified': 132, 'ok': 1.0, '$clusterTime': {'clusterTime': Timestamp(1778587005, 133), 'signature': {'hash': b'\xf0\x83\xad\xe3\xf8\x9c\xca\x0b\x91\x88o\xd4\x99\xe8\x92\xe7N\xb7\xf6\x04', 'keyId': 7592293558351036417}}, 'operationTime': Timestamp(1778587005, 133), 'updatedExisting': True}, acknowledged=True)

**Interpretation**:

This update operation modifies all delivery documents with a "failed" status by assigning a "high" priority value. The output confirms that 132 existing documents were successfully updated in the delivery_cases collection.

### **Update session failure flag**

In [130]:
col_sessions.update_one(
    {"session_id": "S99516"},
    {"$set": {"has_failure": 1}}
)

UpdateResult({'n': 1, 'electionId': ObjectId('7fffffff000000000000014e'), 'opTime': {'ts': Timestamp(1778587006, 7), 't': 334}, 'nModified': 1, 'ok': 1.0, '$clusterTime': {'clusterTime': Timestamp(1778587006, 7), 'signature': {'hash': b"\x98j [\xd2\xa0\xe0A\xc3\xffI[2#\\''\xa3p9", 'keyId': 7592293558351036417}}, 'operationTime': Timestamp(1778587006, 7), 'updatedExisting': True}, acknowledged=True)

**Interpretation:**

This update operation sets the has_failure field to 1 for a specific session (S99516) in the app_sessions collection. The result confirms that one document was successfully modified.

# **Delete**
Removes one or more documents from a MongoDB collection.

### **Delete Session Record**

In [131]:
col_sessions.delete_one(
    {"session_id": "S19847"}
)

DeleteResult({'n': 1, 'electionId': ObjectId('7fffffff000000000000014e'), 'opTime': {'ts': Timestamp(1778587008, 6), 't': 334}, 'ok': 1.0, '$clusterTime': {'clusterTime': Timestamp(1778587008, 6), 'signature': {'hash': b'\xad\xcd\xcc\x18v\xff\x9d\xb9\xf4e\xc9\x0bB\x02\x0e\xfb\x83\x1f\x8d\xc1', 'keyId': 7592293558351036417}}, 'operationTime': Timestamp(1778587008, 6)}, acknowledged=True)

In [132]:
# verify
col_sessions.find_one({"session_id": "S19847"})

**Interpretation**:

This operation removes a single document from the app_sessions collection where session_id is S19847. The result confirms that one matching document was successfully deleted from the database.

**Delete Operation – Data Quality Check**

In [133]:
col_sessions.delete_many({
    "api_latency_ms": {"$type": "string"}
})

DeleteResult({'n': 0, 'electionId': ObjectId('7fffffff000000000000014e'), 'opTime': {'ts': Timestamp(1778587012, 8), 't': 334}, 'ok': 1.0, '$clusterTime': {'clusterTime': Timestamp(1778587012, 8), 'signature': {'hash': b'\x87\xc8\xe4\xdc\x10s~S\x17\xb1\xbd\xa6\xec\xa5PF}\x92\xa3\r', 'keyId': 7592293558351036417}}, 'operationTime': Timestamp(1778587012, 8)}, acknowledged=True)

**Interpretation:**

The output shows n: 0, indicating that no documents matched the deletion condition. This confirms that the dataset was already well-standardised during preprocessing, with no invalid string-type latency values present in the collection.

## **Operational Analysis Using Aggregration Pipeline**

### **Hub-wise failed Deliveries**

In [134]:
result = col_deliveries.aggregate([
    {"$match": {"status": "failed"}},
    {"$group": {
        "_id": "$hub.hub_id",
        "failed_deliveries": {"$sum": 1}
    }},
    {"$sort": {"failed_deliveries": -1}}
])

for doc in result:
    print(doc)

{'_id': 'H08', 'failed_deliveries': 26}
{'_id': 'H05', 'failed_deliveries': 23}
{'_id': 'H01', 'failed_deliveries': 17}
{'_id': 'H04', 'failed_deliveries': 16}
{'_id': 'H06', 'failed_deliveries': 15}
{'_id': 'H07', 'failed_deliveries': 14}
{'_id': 'H03', 'failed_deliveries': 11}
{'_id': 'H02', 'failed_deliveries': 10}


**Interpretation**

The aggregation pipeline grouped failed deliveries by hub location. Hub H08 has the highest number of failed deliveries (26), followed by H05 (23). This indicates that certain hubs experience more operational inefficiencies compared to others. H02 shows the lowest failures (10), suggesting better delivery performance or fewer disruptions in that zone.

### **Customer Complaint Analysis**

In [135]:
pipeline = [
    {"$unwind": "$complaints"},
    {
        "$group": {
            "_id": "$order.customer_id",
            "total_complaints": {"$sum": 1}
        }
    },
    {
        "$sort": {"total_complaints": -1}
    }
]

result = col_deliveries.aggregate(pipeline)

for doc in result:
    print(doc)

{'_id': 'C0368', 'total_complaints': 4}
{'_id': 'C0282', 'total_complaints': 3}
{'_id': 'C0242', 'total_complaints': 3}
{'_id': 'C0421', 'total_complaints': 3}
{'_id': 'C0378', 'total_complaints': 2}
{'_id': 'C0549', 'total_complaints': 2}
{'_id': 'C0561', 'total_complaints': 2}
{'_id': 'C0015', 'total_complaints': 2}
{'_id': 'C0611', 'total_complaints': 2}
{'_id': 'C0208', 'total_complaints': 2}
{'_id': 'C0119', 'total_complaints': 2}
{'_id': 'C0078', 'total_complaints': 2}
{'_id': 'C0508', 'total_complaints': 2}
{'_id': 'C0178', 'total_complaints': 2}
{'_id': 'C0647', 'total_complaints': 2}
{'_id': 'C0578', 'total_complaints': 2}
{'_id': 'C0595', 'total_complaints': 2}
{'_id': 'C0573', 'total_complaints': 2}
{'_id': 'C0517', 'total_complaints': 2}
{'_id': 'C0480', 'total_complaints': 2}
{'_id': 'C0626', 'total_complaints': 2}
{'_id': 'C0001', 'total_complaints': 2}
{'_id': 'C0146', 'total_complaints': 2}
{'_id': 'C0597', 'total_complaints': 2}
{'_id': 'C0571', 'total_complaints': 2}


**Interpretation**

The aggregation results show the number of complaints raised by each customer across all deliveries. Customers such as C0368 (4 complaints) and C0242, C0421 (3 complaints each) represent high-frequency complainants. This indicates repeated service dissatisfaction and highlights customers who may require targeted service improvement or priority resolution handling.

### **Session Performance Analysis**

In [136]:
pipeline = [
    {"$match": {"api_latency_ms": {"$gt": 700}}},
    {
        "$group": {
            "_id": "$device_type",
            "avg_latency": {"$avg": "$api_latency_ms"}
        }
    }
]

for doc in col_sessions.aggregate(pipeline):
    print(doc)

{'_id': 'Android', 'avg_latency': 908.2884615384615}
{'_id': 'iOS', 'avg_latency': 927.5666666666667}
{'_id': 'Web', 'avg_latency': 982.75}


**Interpretation**

Session data was analyzed based on API latency grouped by device type. Web users showed the highest latency, indicating potential performance bottlenecks compared to mobile platforms. The results show that Web sessions have the highest average API latency (≈982 ms), followed by iOS (≈927 ms) and Android (≈908 ms). This indicates that Web-based interactions are comparatively slower in response time, suggesting potential backend inefficiencies or heavier payload handling for web clients.

# **Query Optimisation & Performance Tuning**

Evaluates query performance by comparing execution plans before and after indexing. The objective is to demonstrate how indexing improves data retrieval efficiency by reducing full collection scans.

**Before Indexing**

In [148]:
col_deliveries.find({"status": "failed"}).explain()

{'explainVersion': '1',
 'queryPlanner': {'namespace': 'northstar_db.delivery_cases',
  'parsedQuery': {'status': {'$eq': 'failed'}},
  'indexFilterSet': False,
  'queryHash': '5D6543D9',
  'planCacheShapeHash': '5D6543D9',
  'planCacheKey': '405CB45D',
  'optimizationTimeMillis': 0,
  'maxIndexedOrSolutionsReached': False,
  'maxIndexedAndSolutionsReached': False,
  'maxScansToExplodeReached': False,
  'prunedSimilarIndexes': False,
  'winningPlan': {'isCached': False,
   'stage': 'COLLSCAN',
   'filter': {'status': {'$eq': 'failed'}},
   'direction': 'forward'},
  'rejectedPlans': []},
 'executionStats': {'executionSuccess': True,
  'nReturned': 132,
  'executionTimeMillis': 1,
  'totalKeysExamined': 0,
  'totalDocsExamined': 950,
  'executionStages': {'isCached': False,
   'stage': 'COLLSCAN',
   'filter': {'status': {'$eq': 'failed'}},
   'nReturned': 132,
   'executionTimeMillisEstimate': 0,
   'works': 951,
   'advanced': 132,
   'needTime': 818,
   'needYield': 0,
   'saveState'

**Interpretation:**

Before indexing, MongoDB performed a full collection scan (COLLSCAN) and examined all 950 documents to identify deliveries with status = "failed". This approach is inefficient for large datasets because query performance decreases as data volume increases.

# **Indexing Strategy**

Indexing is used to improve query performance by creating indexes on frequently searched fields. This reduces full collection scans and significantly speeds up filtering and retrieval operations in large datasets.

### **Delivery Cases Indexing**

In [149]:
# Delivery performance indexes
col_deliveries.create_index("status")
col_deliveries.create_index("hub.hub_id")
col_deliveries.create_index("driver.driver_id")
col_deliveries.create_index("order.customer_id")

'order.customer_id_1'

**Interpretation**

Indexes were created on frequently queried fields to improve retrieval speed and reduce full collection scans.

### **Customer Profiles Indexing**

In [150]:
# Customer lookup & analytics indexes
col_customers.create_index("customer_id", unique=True)
col_customers.create_index("home_zone")
col_customers.create_index("engagement.account_status")

'engagement.account_status_1'

**Interpretation**

Indexing improves the performance of customer-related filtering, searching, and analytical queries by enabling faster data retrieval.


### **App Sessions Indexing**

In [151]:
col_sessions.create_index("customer_id")
col_sessions.create_index("event_type")
col_sessions.create_index("api_latency_ms")


'api_latency_ms_1'

**Interpretation**

Indexing improves the performance of session and application analytics queries by enabling faster filtering and retrieval of customer activity data. The indexes created on customer ID, event type, and API latency help optimise user behaviour analysis, event tracking, and application performance monitoring.

## **Query Performance Analysis**

### **After Indexing**

In [152]:
col_deliveries.find({"status": "failed"}).explain()

{'explainVersion': '1',
 'queryPlanner': {'namespace': 'northstar_db.delivery_cases',
  'parsedQuery': {'status': {'$eq': 'failed'}},
  'indexFilterSet': False,
  'queryHash': '5D6543D9',
  'planCacheShapeHash': '5D6543D9',
  'planCacheKey': 'BA4EE208',
  'optimizationTimeMillis': 0,
  'maxIndexedOrSolutionsReached': False,
  'maxIndexedAndSolutionsReached': False,
  'maxScansToExplodeReached': False,
  'prunedSimilarIndexes': False,
  'winningPlan': {'isCached': False,
   'stage': 'FETCH',
   'inputStage': {'stage': 'IXSCAN',
    'keyPattern': {'status': 1},
    'indexName': 'status_1',
    'isMultiKey': False,
    'multiKeyPaths': {'status': []},
    'isUnique': False,
    'isSparse': False,
    'isPartial': False,
    'indexVersion': 2,
    'direction': 'forward',
    'indexBounds': {'status': ['["failed", "failed"]']}}},
  'rejectedPlans': []},
 'executionStats': {'executionSuccess': True,
  'nReturned': 132,
  'executionTimeMillis': 1,
  'totalKeysExamined': 132,
  'totalDocsExami

**Interpretation**

After indexing the status field, MongoDB used an index scan (IXSCAN) and examined only 132 relevant documents instead of scanning the entire collection. This significantly improves query efficiency, reduces processing overhead, and enables faster filtering and analytical operations.

**Justification**

The indexing strategy improves database performance by enabling efficient data retrieval from large collections. Frequently queried fields were indexed based on analytical usage patterns, reducing execution time and eliminating full collection scans for key queries.

# **Schema Validation**
Enforces rules on collections to ensure documents follow a defined structure, data types, and constraints.

## **Delivery Cases**

In [153]:
db.command({
    "collMod": "delivery_cases",
    "validator": {
        "$jsonSchema": {
            "bsonType": "object",
            "required": ["delivery_id", "order_id", "status"],
            "properties": {
                "delivery_id": {"bsonType": "string"},
                "order_id": {"bsonType": "string"},
                "status": {"bsonType": "string"},
                "metrics": {"bsonType": "object"}
            }
        }
    },
    "validationLevel": "moderate"
})

{'ok': 1.0,
 '$clusterTime': {'clusterTime': Timestamp(1778588699, 4),
  'signature': {'hash': b'\xd5\xd4\xb9\xde\xa2\xd4A p\\\xcd\xd3\xd4\n\xf0\x1e7u:\xdb',
   'keyId': 7592293558351036417}},
 'operationTime': Timestamp(1778588699, 4)}

In [154]:
#verify
db.get_collection("delivery_cases").options()

{'validator': {'$jsonSchema': {'bsonType': 'object',
   'required': ['delivery_id', 'order_id', 'status'],
   'properties': {'delivery_id': {'bsonType': 'string'},
    'order_id': {'bsonType': 'string'},
    'status': {'bsonType': 'string'},
    'metrics': {'bsonType': 'object'}}}},
 'validationLevel': 'moderate',
 'validationAction': 'error'}

## **Customer Profile**

In [155]:
db.command({
    "collMod": "customer_profiles",
    "validator": {
        "$jsonSchema": {
            "bsonType": "object",
            "required": ["customer_id"],
            "properties": {
                "customer_id": {"bsonType": "string"},
                "age": {"bsonType": "int"},
                "home_zone": {"bsonType": "string"},
                "engagement": {"bsonType": "object"}
            }
        }
    },
    "validationLevel": "moderate"
})

{'ok': 1.0,
 '$clusterTime': {'clusterTime': Timestamp(1778588705, 1),
  'signature': {'hash': b'\xf5q&\x1f\xdc_\x17\xabU\xb4F;\x9f\xb6jC\x1b\xa6\xf93',
   'keyId': 7592293558351036417}},
 'operationTime': Timestamp(1778588705, 1)}

## **App Sessions**

In [156]:
db.command({
    "collMod": "app_sessions",
    "validator": {
        "$jsonSchema": {
            "bsonType": "object",
            "required": ["session_id", "customer_id"],
            "properties": {
                "session_id": {"bsonType": "string"},
                "customer_id": {"bsonType": "string"},
                "event_type": {"bsonType": "string"},
                "api_latency_ms": {"bsonType": "int"}
            }
        }
    },
    "validationLevel": "moderate"
})

{'ok': 1.0,
 '$clusterTime': {'clusterTime': Timestamp(1778588707, 2),
  'signature': {'hash': b'\xc6a\x12U\xfa\x8d\\\xc6\x16Jsrh\xbd\tM\x02\xf5\x03\xd5',
   'keyId': 7592293558351036417}},
 'operationTime': Timestamp(1778588707, 2)}

**Interpretation:**

Schema validation rules were applied to enforce data consistency.
It enforces a structured format for documents in customer_profiles, delivery_cases, and app_sessions, ensuring that required fields, data types, and nested structures follow the expected schema design. This helps prevent invalid or inconsistent data from being inserted or updated in the database.

**Conclusion**

The MongoDB implementation demonstrates a complete NoSQL lifecycle including schema design, document modeling, CRUD operations, aggregation analysis, and query optimization. The hybrid schema ensures both performance and scalability, while indexing significantly improves query execution efficiency for large-scale operational analytics.